src/Plugin/tmgmt/Translator/ThebigwordTranslator.php

```diff
@@ -0,0 +1,872 @@
+   * @throws \Drupal\tmgmt\TMGMTException
+   */
+  public function uploadFileResource($xliff, JobInterface $job, $tjiid, $name) {
+    $remote_mappings = RemoteMapping::loadByLocalData($job->id(), $tjiid, 'tmgmt_thebigword');
+    $remote_mapping = reset($remote_mappings);
+    $project_id = $remote_mapping->getRemoteIdentifier2();
+
+    /** @var \DateTime $required_by */
+    $required_by = $job->getSetting('required_by');
+    $required_by->setTimezone(new \DateTimeZone('UTC'));
+    $datetime = $required_by->format('Y-m-d\TH:i:s');
+    $form_params = [
+      'ProjectId' => $project_id,
+      'RequiredByDateUtc' => $datetime,
+      'SourceLanguage' => $job->getRemoteSourceLanguage(),
+      'TargetLanguage' => $job->getRemoteTargetLanguage(),
+      'FilePathAndName' => "$name.xliff",
+      'FileState' => 'TranslatableSource',
+      'FileData' => base64_encode($xliff),
+    ];
+    /** @var int $file_id */
+    $file_id = $this->request('file', 'POST', $form_params);
+
+    $files = $remote_mapping->getRemoteData('files');
+    $files[$file_id] = ['FileStateVersion' => 1];
+    $remote_mapping->addRemoteData('files', $files);
+    $remote_mapping->save();
+
+    return $file_id;
+  }
+
+  /**
+   * Parses received translation from thebigword and returns unflatted data.
+   *
+   * @param string $data
+   *   Base64 encode data, received from thebigword.
+   *
+   * @return array
+   *   Unflatted data.
+   */
+  protected function parseTranslationData($data) {
+    /** @var \Drupal\tmgmt_file\Format\FormatInterface $xliff_converter */
+    $xliff_converter = \Drupal::service('plugin.manager.tmgmt_file.format')->createInstance('xlf');
+    // Import given data using XLIFF converter. Specify that passed content is
+    // not a file.
+    return $xliff_converter->import($data, FALSE);
+  }
+
+  /**
+   * Fetches translations for job items of a given job.
+   *
+   * @param \Drupal\tmgmt\JobInterface $job
+   *   A job containing job items that translations will be fetched for.
+   * @param string $state
+   *   The state of the files.
+   *
+   * @return bool
+   *   Returns TRUE if there are error messages during the process of
+   *   retrieving translations. Otherwise FALSE.
+   *
+   * @throws \Drupal\tmgmt\TMGMTException
+   */
+  public function addTranslatedFilesToJob(JobInterface $job, $state) {
+    // Search for placeholder item.
+    $result = RemoteMapping::loadByLocalData($job->id());
+    $remote = reset($result);
+    $this->setTranslator($job->getTranslator());
+    $project_id = $remote->getRemoteIdentifier2();
+    $translated = 0;
+    $not_translated = 0;
+    $had_errors = FALSE;
+
+    try {
+      // Get the files of this job.
+      $files = $this->requestFileinfo($state, $job);
+      /** @var JobItemInterface $job_item */
+      foreach ($job->getItems() as $job_item) {
+        $mappings = RemoteMapping::loadByLocalData($job->id(), $job_item->id(), 'tmgmt_thebigword');
+        $mapping = reset($mappings);
```

Comment:
We need to discuss this method, I don't understand why error handling needs to be so complicated and why you seem to be updating existing files here.